web_drop_target/static/src/js/web_drop_target.js

```diff
@@ -27,44 +28,67 @@ odoo.define('web_drop_target', function(require) {
         },
 
         _on_drop: function(e) {
-            var drop_item = this._get_drop_item(e);
-            if(!drop_item) {
-                return;
-            }
-            jQuery(e.delegateTarget).removeClass(this._drag_over_class);
-            var reader = new FileReader();
-            reader.onloadend = this.proxy(
-                _.partial(this._handle_file_drop, drop_item.getAsFile())
-            );
-            reader.readAsArrayBuffer(drop_item.getAsFile());
+            var self = this;
+            var drop_items = this._get_drop_item(e);
+            _.each(drop_items, function(drop_item) {
+                var reader = new FileReader();
+                reader.onloadend = self.proxy(
+                    _.partial(self._handle_file_drop, drop_item.getAsFile())
+                );
+                reader.readAsArrayBuffer(drop_item.getAsFile());
+            });
+            this._drop_overlay.remove();
+            this._drop_overlay = null;
             e.preventDefault();
         },
 
         _on_dragenter: function(e) {
-            if(this._get_drop_item(e)) {
+            if(this._get_drop_item(e).length) {
                 e.preventDefault();
-                jQuery(e.delegateTarget).addClass(this._drag_over_class);
+                if(!this._drop_overlay){
+                    var drop_overlay_message = _t('Drop your files here');
+                    var o_content = jQuery('.o_content');
+                    var view_manager = jQuery('.o_view_manager_content');
+                    this._drop_overlay = jQuery(
```

Comment:
move this templating to an xml template or split it to separated func to ease override

Reply:
also, use `self.$(...)` to find elements inside the form only

Reply:
I can't change `jQuery(...)` for `self.$` because those classes are outside the Widget, so it wouldn't find anything.
Am I wrong?

Reply:
yes you are. This is a mixin class, so when this is actually instantiated, it does live within a widget: https://github.com/Studio73/web/blob/c33e6fa6745bb6fc59d2ae0f0e681c141e9bbb98/web_drop_target/static/src/js/web_drop_target.js#L138

Reply:
Yes I know that is a mixin class and it does live within a widget. But, what I'm saying is that those classes are in other Widgets, o_content is in ActionManager and o_view_manager_content is in ViewManager. So if you use `this.$('.o_content')` the result it's empty (I've tested it).

Reply:
ah, this way. Well, then you break generality here. It would be better if the addon only blocks/covers the area of the widget (that's also what the drop target is)

Reply:
I've done this because with `position: absolute` when the user scrolls the message stays on top and this is the result:
![3-absolute](https://user-images.githubusercontent.com/4355395/42168232-b3b48814-7e10-11e8-93df-a0f65a29ada1.png)

One solution to this is to use `position: fixed` but, with the fixed property the element is positioned relative to the browser window
![2-fixed](https://user-images.githubusercontent.com/4355395/42168397-294fcc78-7e11-11e8-91f8-cb3868cdfb75.png)

So if we want to cover just the area where the drag&drop works, a possible solution is something like I've done.

AFAIK ActionManager is a property of ViewManager. ViewManager is the parent view of View Widget and FormView, TreeView and KanbanView extends from View. So those css classes are always accesibles and no generality break.

Reply:
concerning generality, think that tomorrow someone wants to use the mixin for enabling some specific action when dropping on some button, div, whatever on a form, but only there. Or a widget in the top bar. Or why not support drag and drop on some menu items that just create some default record based on the dropped file? That's where you break generality.

What's the problem with finding the current widget's DOM element's offset parent, and construct an overlay that covers only that?

Reply:
I've also try to do it in this way but FormView parent `.o_view_manager_content` it's positioned absolute, so `position()` returns `0` and `offset()` changes with the scroll and will became negative at the bottom of the page.  

I'm not opposing to change it, but this is the only way that I found that works.
Honestly, I'm not going to waste more time trying to explain that I already test other solutions. This cover the cases proposed at the README, Form, Tree and Kanban. If someone comes with the idea to support drag&drop in buttons or menus he will be in charge to make all compatible, not me.

Reply:
as far as I see this is not breaking anything, this is ADDING a feature and JUST limiting the generalization. I wouldn't block the change just for this,

That been said, @fuentes010 I know that it can be frustrating to address changes BUT pls don't say that you are wasting time.

Explain what you have done instead (so that ppl don't have to guess what you have tried and what not) and put this clear in the roadmap/known issues session possibly including the **valueable** remark by @hbrunn .

In this way we have a trace and whoever will work again on this widget can address this issue if needed.

Guys pls, remember that we are all on the same side: trying to improve shared code. :wink:

Reply:
@simahawk @hbrunn first of all, thank your for your review.
I didn't mean that I'm against your review and I said cleary that i'm not opposing to change anything, all improvements are welcome, I've change all you asked for. For me address changes it's not frustrating I love to learn. I didn't try to offend to anyone with my comment, if it has been misunderstood I apologise. 

@simahawk I'll change the word waste for invest 😉